Support VFIO passthrough

varunrsekar opened the pull request:
## Design
1. Introduce new featuregate `PassthroughSupport`
2. Discover GPUs using `github.com/NVIDIA/go-nvlib/nvpci` pkg
3. Publish them as vfio devices in the node resource slice of type `vfio` with canonicalName `gpu-vfio-<idx>`
4. Create a new deviceclass `vfio.gpu.nvidia.com`
5. Create a new set of CDI devices in a new file `k8s.gpu.nvidia.com-device_vfio.yaml`:
```
k8s.gpu.nvidia.com/device=gpu-vfio-0
k8s.gpu.nvidia.com/device=gpu-vfio-1
```
6. A successful NodePrepare will cause the device of type `gpu` for the same PCIBusID to be removed from the resourceslice.
7. A successful NodeUnprepare will cause the device of type `gpu` for the same PCIBusID to be rediscovered and added back to the resourceslice

## GPU Discovery
1. Invoke `nvpci.GetGPUs` (`github.com/NVIDIA/go-nvlib/nvpci`)
2. Publish all discovered GPUs as devices of type `vfio` in the resourceslice
3. Set the parent of the vfio device to the device of type `gpu` discovered from NVML (is nil if device is not bound to the nvidia driver)
4. Publish new CDI file (`k8s.gpu.nvidia.com-device_vfio.yaml`) with the discovered vfio devices

## NodePrepare
1. Verify there are no active gpu clients on the device by checking for open FDs against `<driver-root>/dev/nvidia<minor>` (timeout after 60s)
2. Verify there are no VFs on the GPU
3. Configure the GPU for passthrough using introduced `scripts/unbind_from_driver.sh`, `scripts/bind_to_driver.sh` shell scripts.
4. Remove the parent GPU device of type `gpu` in the resourceslice

## NodeUnprepare
1. Configure the GPU back to the nvidia driver using introduced `scripts/unbind_from_driver.sh`, `scripts/bind_to_driver.sh` shell scripts.
2. Rediscover the parent GPU device (as its device minor might've changed) and re-add it back to the resourceslice

## Deployment
1. kubeletplugin is a privileged pod running as root
2. Mount host path `/` (read-only), `/sys/` (read-write) and `/proc/` (read-write)

## Testing
1. Verified with Kubevirt VM using sample resourceclaimtemplate spec from `demo/specs/quickstart/gpu-test-vfiopci.yaml`
2. Regression tests with the specs under https://github.com/NVIDIA/k8s-dra-driver-gpu/tree/main/demo/specs/quickstart/:
    1. With FG enabled: PASS
    2. With FG disabled: PASS

## TODO
1. Use standard attribute name for PCI Bus ID
2. Use nvidia-container-toolkit library to generate the CDI device file (PENDING https://github.com/NVIDIA/nvidia-container-toolkit/pull/315)
3. active GPU VF enablement/disablement
4. Verify there are no VGPU/MIG before advertising VFIO device.

varunrsekar commented on cmd/gpu-kubelet-plugin/vfio-device.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+func execCommandInHostNamespace(cmd string, args []string) ([]byte, error) {
+	nsenterArgs := []string{"--mount=/proc/1/ns/mnt", cmd}
```
Per discussion, I've updated it to use `chroot` to run modprobe from `/run/nvidia/driver`.

shivamerla commented on cmd/gpu-kubelet-plugin/vfio-device.go:
```diff
@@ -0,0 +1,286 @@
+}
+
+func (vm *VfioPciManager) loadVfioPciModule() error {
+	_, err := execCommandWithChroot(vm.driverRoot, "modprobe", []string{vfioPciModule}) //nolint:gosec
```
This assumption doesn’t hold if the driverRoot is not the host root (for example, when the NVIDIA driver is installed at a custom path such as `/home/kubernetes/bin/nvidia`, as on GKE). It’s better to mount the host root explicitly to avoid this issue.

shivamerla commented on cmd/gpu-kubelet-plugin/vfio-device.go:
```diff
@@ -0,0 +1,286 @@
+}
+
+// Configure binds the GPU to the vfio-pci driver.
+func (vm *VfioPciManager) Configure(info *VfioDeviceInfo) error {
```
Pass the context down to all these methods as they can block.

varunrsekar commented on cmd/gpu-kubelet-plugin/vfio-device.go:
```diff
@@ -0,0 +1,286 @@
+	if !vm.isVfioPCIModuleLoaded() {
+		err := vm.loadVfioPciModule()
+		if err != nil {
+			klog.Fatalf("failed to load vfio_pci module: %v", err)
```
We can consider that when we graduate the FG out of Alpha.

varunrsekar commented on cmd/gpu-kubelet-plugin/vfio-device.go:
```diff
@@ -0,0 +1,285 @@
+	if driver == nvidiaDriver {
+		return nil
+	}
+	err = vm.changeDriver(info.pcieBusID, nvidiaDriver)
```
The helm chart needs a lot more work to support running without the `nvidia` driver. For now, I've just included some very limited plumbing for the Configure/Unconfigure calls to do the driver switch differently.

I'll handle further changes incrementally in a follow-up

varunrsekar commented:
Updates:
* Squashed all previous commits
* If MIGs are present, then VFIO devices wouldn't be advertised.
* Updated regression testing in description

varunrsekar commented on pkg/featuregates/featuregates.go:
```diff
@@ -66,6 +69,13 @@ var defaultFeatureGates = map[featuregate.Feature]featuregate.VersionedSpecs{
+		{
+			Default:    false,
+			PreRelease: featuregate.Alpha,
+			Version:    version.MajorMinor(25, 8),
```
Updated it. Thanks

shivamerla commented:
merging this as discussed with @klueska

jgehrcke commented on cmd/gpu-kubelet-plugin/driver.go:
```diff
@@ -158,6 +146,12 @@ func (d *driver) nodePrepareResource(ctx context.Context, claim *resourceapi.Res
 			Err: fmt.Errorf("error preparing devices for claim %v: %w", claim.UID, err),
 		}
 	}
+	if err = d.publishResources(ctx, d.state.config); err != nil {
```
Why are we (re)publishing resources as part of preparing a device?

klueska commented on cmd/gpu-kubelet-plugin/allocatable.go:
```diff
@@ -66,9 +72,79 @@ func (d AllocatableDevice) UUID() string {
 	panic("unexpected type for AllocatableDevice")
 }
 
+type AllocatableDeviceList []*AllocatableDevice
```
Can we drop this?

klueska commented on cmd/gpu-kubelet-plugin/allocatable.go:
```diff
@@ -66,9 +72,79 @@ func (d AllocatableDevice) UUID() string {
 
+type AllocatableDeviceList []*AllocatableDevice
+
+type AllocatableDevices map[string]*AllocatableDevice
```
Can you move this back to the top.

klueska commented on cmd/gpu-kubelet-plugin/allocatable.go:
```diff
@@ -66,9 +72,79 @@ func (d AllocatableDevice) UUID() string {
+
+type AllocatableDevices map[string]*AllocatableDevice
+
+func (d AllocatableDevices) getDevicesByGPUPCIBusID(pcieBusID string) AllocatableDeviceList {
```
Can we group the functions together that are only here because we dont yet support this feature via PartitionableDevices? Then have a comment before that block saying we can remove these once we support that.

klueska commented on cmd/gpu-kubelet-plugin/cdi.go:
```diff
@@ -139,7 +142,79 @@ func NewCDIHandler(opts ...cdiOption) (*CDIHandler, error) {
 	return h, nil
 }
 
+func (cdi *CDIHandler) writeSpec(spec spec.Interface, specName string) error {
```
Can we mvoe this to the bottom of the file. We generally prefer to have private methods below public ones.

varunrsekar commented on scripts/bind_to_driver.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
```
@cdesiniotis Thanks. the library is very close to what we need here. Should this be migrated to NVIDIA/go-nvlib so that it can be consumed in both places?
